packages/babel-register/test/async.js

```diff
@@ -0,0 +1,252 @@
+    // already finished.
+    await new Promise(setImmediate);
+
+    revertRegister();
+    currentHook = null;
+    currentOptions = null;
+    sourceMapSupport = false;
+    jest.resetModules();
+  });
+
+  afterAll(() => {
+    resetCache();
+  });
+
+  test("registers hook correctly", () => {
+    setupRegister();
+
+    expect(typeof currentHook).toBe("function");
+    expect(currentOptions).toEqual(defaultOptions);
+  });
+
+  test("unregisters hook correctly", () => {
+    setupRegister();
+    revertRegister();
+
+    expect(currentHook).toBeNull();
+    expect(currentOptions).toBeNull();
+  });
+
+  test("installs source map support by default", () => {
+    setupRegister();
+
+    currentHook("const a = 1;", testFile);
+
+    expect(sourceMapSupport).toBe(true);
+  });
+
+  test("installs source map support when requested", () => {
+    setupRegister({
+      babelrc: false,
+      sourceMaps: true,
+    });
+
+    currentHook("const a = 1;", testFile);
+
+    expect(sourceMapSupport).toBe(true);
+  });
+
+  test("does not install source map support if asked not to", () => {
+    setupRegister({
+      babelrc: false,
+      sourceMaps: false,
+    });
+
+    currentHook("const a = 1;", testFile);
+
+    expect(sourceMapSupport).toBe(false);
+  });
+
+  it("returns concatenatable sourceRoot and sources", callback => {
+    // The Source Maps R3 standard https://sourcemaps.info/spec.html states
+    // that `sourceRoot` is “prepended to the individual entries in the
+    // ‘source’ field.” If `sources` contains file names, and `sourceRoot`
+    // is intended to refer to a directory but doesn’t end with a trailing
+    // slash, any consumers of the source map are in for a bad day.
+    //
+    // The underlying problem seems to only get triggered if one file
+    // requires() another with @babel/register active, and I couldn’t get
+    // that working inside a test, possibly because of jest’s mocking
+    // hooks, so we spawn a separate process.
+
+    spawnNode(["-r", registerFile, sourceMapTestFile], output => {
+      let err;
+
+      try {
+        const sourceMap = JSON.parse(output);
+        expect(sourceMap.map.sourceRoot + sourceMap.map.sources[0]).toBe(
+          sourceMapNestedTestFile,
+        );
+      } catch (e) {
+        err = e;
+      }
+
+      callback(err);
+    });
+  });
+
+  test("hook transpiles with config", () => {
+    setupRegister({
+      babelrc: false,
+      sourceMaps: false,
+      plugins: ["@babel/transform-modules-commonjs"],
+    });
+
+    const result = currentHook(testFileContent, testFile);
+
+    expect(result).toBe('"use strict";\n\nrequire("assert");');
+  });
+
+  test("hook transpiles with babelrc", () => {
+    setupRegister({
+      babelrc: true,
+      sourceMaps: false,
+    });
+
+    const result = currentHook(testFileContent, testFile);
+
+    expect(result).toBe('"use strict";\n\nrequire("assert");');
+  });
+
+  test("hook supports .mjs config files", () => {
```

Comment:
Except for this test, this file is a copy of https://github.com/babel/babel/blob/main/packages/babel-register/test/index.js